File: app/src/test/java/de/enbeplus/whopper/OpeningHoursTest.kt
package de.enbeplus.whopper

import de.enbeplus.whopper.model.OpenState
import de.enbeplus.whopper.model.OpeningHours
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.temporal.TemporalAdjusters

/**
 * Alle Zeitangaben in diesen Tests stammen wortwoertlich aus OSM, aus der Stichprobe
 * entlang der Strecke Dortmund - Stuttgart.
 */
class OpeningHoursTest {

    private fun at(day: DayOfWeek, hour: Int, minute: Int = 0): LocalDateTime =
        LocalDate.of(2026, 6, 1)
            .with(TemporalAdjusters.nextOrSame(day))
            .atTime(hour, minute)

    private fun open(spec: String, moment: LocalDateTime): OpenState.Open {
        val state = OpeningHours.evaluate(spec, moment)
        assertTrue("erwartet offen, war $state", state is OpenState.Open)
        return state as OpenState.Open
    }

    private fun closed(spec: String, moment: LocalDateTime): OpenState.Closed {
        val state = OpeningHours.evaluate(spec, moment)
        assertTrue("erwartet geschlossen, war $state", state is OpenState.Closed)
        return state as OpenState.Closed
    }

    @Test
    fun `24 durch 7 ist immer offen`() {
        val state = open("24/7", at(DayOfWeek.SUNDAY, 3, 30))
        assertNull(state.until)
    }

    @Test
    fun `fenster ueber mitternacht reicht in den folgetag`() {
        val spec = "Mo-Th 09:00-01:00, Fr,Sa 09:00-03:00, Su 10:00-01:00"

        // Mittwoch 23 Uhr: offen, und zwar bis Donnerstag 01:00.
        val wednesday = open(spec, at(DayOfWeek.WEDNESDAY, 23))
        assertEquals(DayOfWeek.THURSDAY, wednesday.until!!.dayOfWeek)
        assertEquals(LocalTime.of(1, 0), wednesday.until!!.toLocalTime())

        // Samstag 02 Uhr faellt noch in das Freitagsfenster bis 03:00.
        val saturdayNight = open(spec, at(DayOfWeek.SATURDAY, 2))
        assertEquals(LocalTime.of(3, 0), saturdayNight.until!!.toLocalTime())

        // Donnerstag 02 Uhr ist zu, geoeffnet wird erst wieder um 09:00.
        val thursdayEarly = closed(spec, at(DayOfWeek.THURSDAY, 2))
        assertEquals(DayOfWeek.THURSDAY, thursdayEarly.nextOpen!!.dayOfWeek)
        assertEquals(LocalTime.of(9, 0), thursdayEarly.nextOpen!!.toLocalTime())
    }

    @Test
    fun `zwei fenster am tag werden ueber mitternacht zusammengefasst`() {
        val spec = "Mo-Fr 00:00-01:00,10:00-24:00; Sa,Su 00:00-03:00,10:00-24:00"

        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.MONDAY, 0, 30)) is OpenState.Open)
        val gap = closed(spec, at(DayOfWeek.MONDAY, 9))
        assertEquals(LocalTime.of(10, 0), gap.nextOpen!!.toLocalTime())

        // Montag 23 Uhr laeuft nahtlos in das Dienstagsfenster bis 01:00 weiter.
        val evening = open(spec, at(DayOfWeek.MONDAY, 23))
        assertEquals(DayOfWeek.TUESDAY, evening.until!!.dayOfWeek)
        assertEquals(LocalTime.of(1, 0), evening.until!!.toLocalTime())
    }

    @Test
    fun `wochentagsbereich ueber den sonntag hinweg`() {
        val spec = "Su-Th 11:00-01:00, Fr,Sa 11:00-02:00"
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.SUNDAY, 12)) is OpenState.Open)
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.TUESDAY, 12)) is OpenState.Open)
        val saturday = open(spec, at(DayOfWeek.SATURDAY, 23))
        assertEquals(LocalTime.of(2, 0), saturday.until!!.toLocalTime())
    }

    @Test
    fun `regel ohne wochentag gilt jeden tag`() {
        val spec = "05:30-22:30"
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.SUNDAY, 6)) is OpenState.Open)
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.WEDNESDAY, 23)) is OpenState.Closed)
    }

    @Test
    fun `feiertage werden ignoriert und das ergebnis markiert`() {
        val state = open("Mo-Su,PH 11:00-22:00", at(DayOfWeek.SUNDAY, 12))
        assertTrue(state.holidaysIgnored)
        assertFalse(open("Mo-Su 11:00-22:00", at(DayOfWeek.SUNDAY, 12)).holidaysIgnored)
    }

    @Test
    fun `semikolon und komma als regeltrenner gemischt`() {
        val spec = "Mo-Th 10:00-22:00; Fr 10:00-23:00, Sa 10:00-23:00, Su 10:00-23:00, PH 10:00-23:00"
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.MONDAY, 22, 30)) is OpenState.Closed)
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.FRIDAY, 22, 30)) is OpenState.Open)
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.SUNDAY, 22, 30)) is OpenState.Open)
    }

    @Test
    fun `off-regel schliesst einen einzelnen tag`() {
        val spec = "Mo-Su 10:00-20:00; We off"
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.TUESDAY, 12)) is OpenState.Open)
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.WEDNESDAY, 12)) is OpenState.Closed)
    }

    @Test
    fun `zerstueckelte echte angabe bleibt auswertbar`() {
        // So steht es tatsaechlich in OSM, inklusive der doppelten Nennung von Mo.
        val spec = "Mo 00:00-01:00,08:00-01:00; Tu-Th 08:00-01:00; Fr 08:00-24:00; Sa,Su 00:00-24:00"
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.SATURDAY, 4)) is OpenState.Open)
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.TUESDAY, 7)) is OpenState.Closed)
        assertTrue(OpeningHours.evaluate(spec, at(DayOfWeek.TUESDAY, 9)) is OpenState.Open)
    }

    @Test
    fun `was nicht sicher lesbar ist wird als unbekannt gemeldet statt geraten`() {
        listOf(
            null,
            "",
            "Mar-Oct 10:00-20:00",
            "sunrise-sunset",
            "Mo[1] 10:00-20:00",
            "Dec 24 10:00-14:00",
        ).forEach { spec ->
            val state = OpeningHours.evaluate(spec, at(DayOfWeek.MONDAY, 12))
            assertTrue("$spec ergab $state", state is OpenState.Unknown)
        }
    }
}

File: tests/opening-hours.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { evaluateOpeningHours } from '../web/opening-hours.js';

// Fester Bezugstag, damit die Wochentage in den Tests eindeutig sind.
// 2026-06-01 ist ein Montag.
const MONDAY = new Date(2026, 5, 1);

function at(dayOffset, hour, minute = 0) {
  const moment = new Date(MONDAY);
  moment.setDate(MONDAY.getDate() + dayOffset);
  moment.setHours(hour, minute, 0, 0);
  return moment;
}

const MO = 0, TU = 1, WE = 2, TH = 3, FR = 4, SA = 5, SU = 6;

test('Bezugstag ist wirklich ein Montag', () => {
  assert.equal(MONDAY.getDay(), 1);
});

test('24/7 ist immer offen', () => {
  const state = evaluateOpeningHours('24/7', at(SU, 3, 30));
  assert.equal(state.state, 'open');
  assert.equal(state.until, null);
});

test('Fenster ueber Mitternacht reicht in den Folgetag', () => {
  const spec = 'Mo-Th 09:00-01:00, Fr,Sa 09:00-03:00, Su 10:00-01:00';

  const wednesday = evaluateOpeningHours(spec, at(WE, 23));
  assert.equal(wednesday.state, 'open');
  assert.equal(wednesday.until.getHours(), 1);
  assert.equal(wednesday.until.getDate(), at(TH, 1).getDate());

  const saturdayNight = evaluateOpeningHours(spec, at(SA, 2));
  assert.equal(saturdayNight.state, 'open');
  assert.equal(saturdayNight.until.getHours(), 3);

  const thursdayEarly = evaluateOpeningHours(spec, at(TH, 2));
  assert.equal(thursdayEarly.state, 'closed');
  assert.equal(thursdayEarly.nextOpen.getHours(), 9);
});

test('Zwei Fenster am Tag werden ueber Mitternacht zusammengefasst', () => {
  const spec = 'Mo-Fr 00:00-01:00,10:00-24:00; Sa,Su 00:00-03:00,10:00-24:00';
  assert.equal(evaluateOpeningHours(spec, at(MO, 0, 30)).state, 'open');

  const gap = evaluateOpeningHours(spec, at(MO, 9));
  assert.equal(gap.state, 'closed');
  assert.equal(gap.nextOpen.getHours(), 10);

  const evening = evaluateOpeningHours(spec, at(MO, 23));
  assert.equal(evening.state, 'open');
  assert.equal(evening.until.getHours(), 1);
});

test('Wochentagsbereich ueber den Sonntag hinweg', () => {
  const spec = 'Su-Th 11:00-01:00, Fr,Sa 11:00-02:00';
  assert.equal(evaluateOpeningHours(spec, at(SU, 12)).state, 'open');
  assert.equal(evaluateOpeningHours(spec, at(TU, 12)).state, 'open');
  assert.equal(evaluateOpeningHours(spec, at(SA, 23)).until.getHours(), 2);
});

test('Regel ohne Wochentag gilt jeden Tag', () => {
  assert.equal(evaluateOpeningHours('05:30-22:30', at(SU, 6)).state, 'open');
  assert.equal(evaluateOpeningHours('05:30-22:30', at(WE, 23)).state, 'closed');
});

test('Feiertagsregeln werden ignoriert und markiert', () => {
  const state = evaluateOpeningHours('Mo-Su,PH 11:00-22:00', at(SU, 12));
  assert.equal(state.state, 'open');
  assert.equal(state.holidaysIgnored, true);
  assert.equal(evaluateOpeningHours('Mo-Su 11:00-22:00', at(SU, 12)).holidaysIgnored, false);
});

test('Eine reine PH-Regel gilt nicht an Wochentagen', () => {
  // Der Fehler, den der Android-Test gefunden hatte: PH 10:00-23:00 galt taeglich.
  const spec = 'Mo-Th 10:00-22:00; Fr 10:00-23:00, Sa 10:00-23:00, Su 10:00-23:00, PH 10:00-23:00';
  assert.equal(evaluateOpeningHours(spec, at(MO, 22, 30)).state, 'closed');
  assert.equal(evaluateOpeningHours(spec, at(FR, 22, 30)).state, 'open');
});

test('off-Regel schliesst einen einzelnen Tag', () => {
  const spec = 'Mo-Su 10:00-20:00; We off';
  assert.equal(evaluateOpeningHours(spec, at(TU, 12)).state, 'open');
  assert.equal(evaluateOpeningHours(spec, at(WE, 12)).state, 'closed');
});

test('Nicht sicher Lesbares wird als unbekannt gemeldet', () => {
  for (const spec of [null, '', 'Mar-Oct 10:00-20:00', 'sunrise-sunset', 'Mo[1] 10:00-20:00']) {
    assert.equal(evaluateOpeningHours(spec, at(MO, 12)).state, 'unknown', String(spec));
  }
});

// Alle unterschiedlichen Angaben, die in der Stichprobe tatsaechlich in OSM stehen.
const REAL = [
  "00:00-01:00,10:00-24:00",
  "05:30-22:30",
  "08:00-22:00",
  "10:30-00:30; Fr,Sa 09:30-02:30",
  "Mo 00:00-01:00,08:00-01:00; Tu-Th 08:00-01:00; Fr 08:00-24:00; Sa,Su 00:00-24:00",
  "Mo 10:00-01:00, Tu 10:00-01:00, We 10:00-01:00, Th 10:00-01:00, Fr 10:00-02:00, Sa 10:00-02:00, Su 11:00-01:00",
  "Mo,Th-Su 06:00-22:00; Tu-We 08:00-20:00",
  "Mo-Fr 00:00-01:00,10:00-24:00; Sa,Su 00:00-03:00,10:00-24:00",
  "Mo-Fr 06:30-22:00; Sa 06:30-21:00; Su 06:30-22:00; PH 06:30-22:00",
  "Mo-Fr 10:00-22:00, Sa 10:00-22:00, Su 10:00-22:00, PH 10:00-22:00",
  "Mo-Fr 10:30-24:00, Sa,Su 11:00-24:00",
  "Mo-Sa 08:00-24:00",
  "Mo-Sa 10:00-24:00, Su 11:00-24:00",
  "Mo-Sa 12:00-20:00; Su 12:00-22:00",
  "Mo-Su 09:00-24:00",
  "Mo-Su 10:00-24:00",
  "Mo-Su 11:30-00:00",
  "Mo-Su,PH 10:00-16:30",
  "Mo-Su,PH 11:00-22:00",
  "Mo-Th 08:00-01:00, Fr 08:00-03:30, Sa 08:00-04:00, Su 10:00-01:00",
  "Mo-Th 08:00-01:00, Fr 08:00-05:00, Sa 08:00-05:00, Su 09:00-24:00, PH 09:00-24:00",
  "Mo-Th 08:00-01:00, Fr,Sa 08:00-03:00, Su 10:00-01:00",
  "Mo-Th 09:00-01:00, Fr 09:00-02:00, Sa 09:00-02:00, Su 10:00-01:00, PH 10:00-01:00",
  "Mo-Th 09:00-01:00, Fr,Sa 09:00-03:00, Su 10:00-01:00",
  "Mo-Th 09:00-24:00; Fr 00:00-01:00,09:00-24:00; Sa,Su 00:00-03:00,10:00-24:00",
  "Mo-Th 09:00-24:00; Fr 09:00-03:00; Sa 09:00-03:00; Su 10:00-24:00",
  "Mo-Th 09:00-24:00; Tu-Fr 00:00-01:00; Fr,Sa 09:00-24:00; Sa,Su 00:00-03:00; Su 10:00-24:00; Mo 00:00-01:00",
  "Mo-Th 09:30-02:00, Fr,Sa 10:00-04:00, Su 11:00-24:00",
  "Mo-Th 10:00-00:00; Fr,Sa 10:00-02:00; Su 10:00-24:00",
  "Mo-Th 10:00-01:00, Fr 10:00-03:00, Sa 10:00-02:00, Su 10:00-01:00, PH 10:00-01:00",
  "Mo-Th 10:00-01:00, Fr 10:00-03:00, Sa 10:00-03:00, Su 10:00-01:00, PH 10:00-01:00",
  "Mo-Th 10:00-01:00, Fr 10:00-03:00, Sa 10:00-03:00, Su 10:00-24:00, PH 10:00-24:00",
  "Mo-Th 10:00-22:00; Fr 10:00-23:00, Sa 10:00-23:00, Su 10:00-23:00, PH 10:00-23:00",
  "Mo-Th 10:00-22:00; Fr,Sa 11:00-24:00; Su 12:00-22:00",
  "Mo-Th 10:00-23:00; Fr-Sa 10:00-24:00; Su 10:30-23:00",
  "Mo-Th 10:00-24:00, Fr 10:00-01:00, Sa 10:00-01:00, Su 10:00-24:00, PH 10:00-24:00",
  "Mo-Th 10:00-24:00, Fr 10:00-02:00, Sa 10:00-02:00, Su 10:00-24:00, PH 10:00-24:00",
  "Mo-Th 10:00-24:00, Fr 10:00-03:00, Sa 10:00-03:00, Su 10:00-24:00, PH 10:00-24:00",
  "Mo-Th 10:00-24:00, Su 10:30-24:00, PH 10:00-24:00, Fr 10:00-02:00, Sa 10:30-02:00",
  "Mo-Th 11:00-01:00, Fr-Su 11:00-02:00",
  "Mo-Th 11:00-22:00; Fr 11:00-03:00; Sa 11:00-05:00; Su 00:00-24:00",
  "Mo-Th 11:00-23:00, Fr 11:00-24:00, Sa 11:00-24:00, Su 11:00-23:00, PH 11:00-23:00",
  "Mo-Th 11:00-23:00, Fr,Sa 11:00-24:00, Su 11:00-23:00",
  "Mo-Th 11:30-21:30; Fr-Su 11:30-22:00",
  "Mo-Th 12:00-20:00; Fr,Sa 11:00-21:00; Su 11:00-20:00",
  "Mo-Th 12:00-21:00, Fr 12:00-22:00, Sa 12:00-22:00, Su 12:00-21:00, PH 12:00-21:00",
  "PH,Mo-Th 10:00-24:00; Fr,Sa 10:00-03:00",
  "Su 11:00-01:00, Mo-Th 10:00-01:00, Fr,Sa 10:00-02:00",
  "Su-Th 09:00-24:00, Fr,Sa 09:00-01:00",
  "Su-Th 10:00-01:00, Fr,Sa 10:00-03:00",
  "Su-Th 10:30-01:00, Fr 10:30-03:00, Sa 11:00-03:00",
  "Su-Th 11:00-01:00, Fr,Sa 11:00-02:00",
  "Su-Th 11:00-24:00, Fr,Sa 11:00-02:00",
  "Su-We 10:00-01:00, Th 10:00-02:00, Fr,Sa 10:00-04:00",
];

test('Jede echte Angabe ist auswertbar', () => {
  for (const spec of REAL) {
    assert.notEqual(evaluateOpeningHours(spec, at(WE, 12)).state, 'unknown', spec);
  }
});

test('Jede echte Angabe hat offene und geschlossene Zeiten in der Woche', () => {
  for (const spec of REAL) {
    const states = new Set();
    for (let half = 0; half < 7 * 24 * 2; half += 1) {
      const moment = new Date(MONDAY.getTime() + half * 30 * 60_000);
      states.add(evaluateOpeningHours(spec, moment).state);
    }
    assert.ok(states.has('open'), `nie offen: ${spec}`);
    assert.ok(states.has('closed'), `nie geschlossen: ${spec}`);
  }
});
